bin/onlyswaps-smoketest/src/smoketest.rs

```diff
@@ -0,0 +1,155 @@
+            do_swap(&client, &fee_estimator, &swap, req_builder.clone())
+                .instrument(span)
+                .await;
+
+            interval.tick().await;
+        }
+    })
+}
+
+/// Fetch the latest solver / network fee, and execute a swap
+async fn do_swap(
+    client: &Arc<OnlySwapsClient>,
+    fee_estimator: &FeeEstimator,
+    swap: &SwapTest,
+    req_builder: OnlySwapsRequestBuilder,
+) {
+    // Fetch the latest fees from the api
+    let req_with_fees = match req_builder
+        .clone()
+        .exact_amount(swap.amount, fee_estimator)
+        .await
+    {
+        Ok(req_with_fees) => req_with_fees,
+        Err(e) => {
+            tracing::error!(error = ?e, "failed to fetch fees");
+            return;
+        }
+    };
+
+    let req = req_with_fees.build().expect("valid arguments"); // only fallible if misused
+
+    // Attempt to send a new swap request
+    let swap_res = client.approve_and_swap(req).await;
+    let receipt = match swap_res {
+        Ok(receipt) => receipt,
+        Err(e) => {
+            tracing::error!(error = ?e, "failed to request swap through client");
+            return;
+        }
+    };
+
+    // We have successfully submitted a swap
+    let span = tracing::warn_span!("swap_submitted", request_id = %receipt.request_id, tx_hash = %receipt.tx_hash).entered();
+    tracing::info!("Swap request submitted");
+    Metrics::report_swap_requested(swap.label.clone());
+
+    // Monitor swap
+    monitor_swap(client, swap, &receipt)
+        .instrument(span.exit()) // exit the EnteredSpan to get a span that can be used to instrument fut
+        .await
+}
+
+/// Monitor a swap receipt and log stages
+async fn monitor_swap(client: &Arc<OnlySwapsClient>, swap: &SwapTest, receipt: &OnlySwapsReceipt) {
+    // Checks that the swap is fulfilled & verified within the timeout
+    let timeout_res = tokio::time::timeout(swap.timeout, async {
+        // First, wait for completion
+        tracing::debug!("Waiting for swap completion");
+        match client.wait_until_complete(receipt).await {
+            Ok(()) => {
+                tracing::info!("Swap request fulfilled");
+                Metrics::report_swap_fulfilled(swap.label.clone());
+            }
+            Err(e) => {
+                tracing::error!(error = ?e, "Failed to wait until swap is fulfilled");
+                return;
+            }
+        }
+
+        // Then, for verification
+        tracing::debug!("Waiting for swap verification");
+        match client.wait_until_verified(receipt).await {
+            Ok(()) => {
+                tracing::info!("Swap request verified");
+                Metrics::report_swap_verified(swap.label.clone());
+            }
+            Err(e) => {
+                tracing::error!(error = ?e, "Failed to wait until swap is verified");
+            }
+        }
+    })
+    .await;
+
+    if timeout_res.is_err() {
```

Comment:
and could also consider `inspect_err` here

Reply:
i ended up using a match arm to report metrics for client error & timeout